android-core/src/main/kotlin/com/mparticle/internal/messages/BaseMPMessageBuilder.kt

```diff
@@ -0,0 +1,132 @@
+            put(Constants.MessageKey.NAME, name)
+        } catch (e: JSONException) {
+            e.printStackTrace()
+        }
+        return this
+    }
+
+    fun attributes(attributes: JSONObject?): BaseMPMessageBuilder {
+        if (attributes != null && attributes.length() > 0) {
+            try {
+                put(Constants.MessageKey.ATTRIBUTES, attributes)
+                mLength?.let { addEventLengthAttributes(it) }
+            } catch (e: JSONException) {
+                e.printStackTrace()
+            }
+        }
+        return this
+    }
+
+    val messageType: String
+        get() = optString(Constants.MessageKey.TYPE)
+
+    @Throws(JSONException::class)
+    open fun build(session: InternalSession, location: Location?, mpId: Long): BaseMPMessage {
+        return BaseMPMessage(this, session, location, mpId)
+    }
+
+    fun dataConnection(dataConnection: String): BaseMPMessageBuilder {
+        try {
+            put(Constants.MessageKey.STATE_INFO_DATA_CONNECTION, dataConnection)
+        } catch (e: JSONException) {
+            e.printStackTrace()
+        }
+        return this
+    }
+
+    fun length(length: Double?): BaseMPMessageBuilder {
+        mLength = length
+        if (length != null) {
+            try {
+                put(Constants.MessageKey.EVENT_DURATION, length)
+                addEventLengthAttributes(length)
+            } catch (e: JSONException) {
+                e.printStackTrace()
+            }
+        }
+        return this
+    }
+
+    private fun addEventLengthAttributes(length: Double) {
+        try {
+            if (!has(Constants.MessageKey.ATTRIBUTES)) {
+                put(Constants.MessageKey.ATTRIBUTES, JSONObject())
+            }
+            if (!getJSONObject(Constants.MessageKey.ATTRIBUTES).has("EventLength")) {
+                //can't be longer than max int milliseconds
+                getJSONObject(Constants.MessageKey.ATTRIBUTES).put("EventLength", length.toInt().toString())
+            }
+        } catch (e: JSONException) {
+            e.printStackTrace()
+        }
+    }
+
+    fun flags(customFlags: Map<String?, List<String?>>?): BaseMPMessageBuilder {
+        if (customFlags != null) {
+            try {
+                val flagsObject = JSONObject()
+                for ((key, values) in customFlags) {
+                    val valueArray = JSONArray(values)
+                    flagsObject.put(key, valueArray)
+                }
+                put(Constants.MessageKey.EVENT_FLAGS, flagsObject)
+            } catch (e: JSONException) {
+                e.printStackTrace()
+            }
+        }
+        return this
+    }
+
+    val keys: Array<String?>
+        get() {
+            val strings: MutableList<String?> = ArrayList<String?>()
+            val iterator = keys()
+            while (iterator.hasNext()) {
+                strings.add(iterator.next())
+            }
+            return strings.toTypedArray<String?>()
+        }
```

Comment:
We should try to use functional style in cases like this as it's much more concise (please confirm the syntax in my suggestion actually works correctly and double check performance before committing the change, but if performance is similar, I think the functional style is much cleaner and easier to read.

```suggestion
    val keys: Array<String?>
        get() = keys().asSequence().toList().toTypedArray()
```